datastore/complete_post_object.go

```diff
@@ -0,0 +1,86 @@
+package datastore
+
+import (
+	"fmt"
+
+	nex "github.com/PretendoNetwork/nex-go"
+	common_globals "github.com/PretendoNetwork/nex-protocols-common-go/globals"
+	datastore "github.com/PretendoNetwork/nex-protocols-go/datastore"
+	datastore_types "github.com/PretendoNetwork/nex-protocols-go/datastore/types"
+)
+
+func completePostObject(err error, client *nex.Client, callID uint32, param *datastore_types.DataStoreCompletePostParam) uint32 {
+	if commonDataStoreProtocol.s3ObjectSizeHandler == nil {
+		common_globals.Logger.Warning("S3ObjectSize not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.getObjectSizeDataIDHandler == nil {
+		common_globals.Logger.Warning("GetObjectSizeDataID not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.updateObjectUploadCompletedByDataIDHandler == nil {
+		common_globals.Logger.Warning("UpdateObjectUploadCompletedByDataID not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return nex.Errors.DataStore.Unknown
+	}
+
+	// TODO - What is param.IsSuccess? Is this correct?
+	if param.IsSuccess {
+		bucket := commonDataStoreProtocol.s3Bucket
+		key := fmt.Sprintf("%d.bin", param.DataID)
```

Comment:
I'm not versed in how HPP operates, why does it need it's own folder? I'm not opposed to this change (SMM originally used the folder `10.WUP_AMAJ_datastore/ds/1/data`, so it's not unheard of), I just wanna make sure I understand the usecase

Reply:
HPP has two subdirectories for handling S3 files: `data` and `notify` (with the main directory being the same `10.WUP_KCFJ_datastore/ds/1`).

The data folder has the meta binary that the client uploads to S3. Then, there is a `notify` folder which the client uses to determine if it has received a new "notification", which would be if someone uploads a meta binary and the client has permission to access it.

The client accesses an S3 notification URL identified with its PID such as `notify/{%011d}`, without the brackets `{}`. The URL is provided by the HPP server. The contents of a notification can be seen [here](https://github.com/PretendoNetwork/pokemon-rumble-world-secure/blob/e6e7aa8a239bbc9743c8866b244344f95ffb3259/nex/datastore/complete_post_object_v1.go#L51).

We would split the data and notifications to avoid potential collisions.

Reply:
> HPP has two subdirectories for handling S3 files: `data` and `notify` (with the main directory being the same `10.WUP_KCFJ_datastore/ds/1`).

Okay, I think I'll actually just mimic this structure entirely tbh. I'm going to assume that 3DS titles will use `CTR` instead of `WUP`



> The data folder has the meta binary that the client uploads to S3

I'm assuming this is just what this specific game decides to do, yes? Because in SMM and Tipping Stars this is not the case, the data uploaded to S3 does not match the meta binary. Or does HPP just work completely differently here?

> Then, there is a `notify` folder which the client uses to determine if it has received a new "notification", which would be if someone uploads a meta binary and the client has permission to access it.

Gotcha. This is in stark contrast to how SMM handles notifications since it uses BufferQueue's and object ratings for that (and doesn't use HPP)

> The client accesses an S3 notification URL identified with its PID such as `notify/{%011d}`, without the brackets `{}`. The URL is provided by the HPP server. The contents of a notification can be seen [here](https://github.com/PretendoNetwork/pokemon-rumble-world-secure/blob/e6e7aa8a239bbc9743c8866b244344f95ffb3259/nex/datastore/complete_post_object_v1.go#L51).
> 
> We would split the data and notifications to avoid potential collisions.

So the `notify` content will *ALWAYS* be the users PID? Is this hard coded into the library or does the server provide the *upload* URL (I see you mentioned the accessing URL, but didn't mention uploading)? I agree with splitting, I just want to make sure that the implementation is correct. Right now whenever the client wants to upload an object or a meta binary it creates a new DataStore object for the upload

Are these `notify` files objects as well? Do they get DataIDs like regular objects? Or do we now have to maintain separate tables for standard objects vs notifications?

Also a note on the notification contents you linked

> // TODO - What is the last number? Looks constant per-game

`1425562994` is the unix timestamp for March 5, 2015, which is VERY close to the release date of Rumble World (April 8, 2015). Could very likely be related?

Reply:
> Okay, I think I'll actually just mimic this structure entirely tbh. I'm going to assume that 3DS titles will use `CTR` instead of `WUP`

Yes, it's `CTR`. I had copied your SMM example and forgot to rename that 😅

> I'm assuming this is just what this specific game decides to do, yes? Because in SMM and Tipping Stars this is not the case, the data uploaded to S3 does not match the meta binary. Or does HPP just work completely differently here?

Sorry, I was mixing things up. Yes, this is the data that gets uploaded, the meta binary is empty aswell.


> So the `notify` content will _ALWAYS_ be the users PID? Is this hard coded into the library or does the server provide the _upload_ URL (I see you mentioned the accessing URL, but didn't mention uploading)? I agree with splitting, I just want to make sure that the implementation is correct. Right now whenever the client wants to upload an object or a meta binary it creates a new DataStore object for the upload

The S3 key for notifications is the user's PID, just like the S3 key for normal data is the Data ID. The notification is uploaded by the HPP server itself.

> Are these `notify` files objects as well? Do they get DataIDs like regular objects? Or do we now have to maintain separate tables for standard objects vs notifications?
> 
> Also a note on the notification contents you linked
> 
> > // TODO - What is the last number? Looks constant per-game
> 
> `1425562994` is the unix timestamp for March 5, 2015, which is VERY close to the release date of Rumble World (April 8, 2015). Could very likely be related?

This actually could make sense. I checked the Swapdoodle value and it gives the timestamp on November 14 2016, a few days before the first release on (November 17, 2016)

Reply:
> The notification is uploaded by the HPP server itself

Gotcha gotcha, so we don't have to worry about creating objects in the database for this then. For all intents and purposes it sounds very "internal" so I think we can just keep this separate from the normal objects database. Good to know

Reply:
This should be resolved now. The server now takes in a base for the key used for both data and notify contents